src/main/java/build/buildfarm/server/ByteStreamService.java

```diff
@@ -106,9 +106,20 @@ private void readBlob(
       return;
     }
 
-    responseObserver.onNext(ReadResponse.newBuilder()
-        .setData(blob)
-        .build());
+    while (!blob.isEmpty()) {
+      ByteString chunk;
+      if (blob.size() < DEFAULT_CHUNK_SIZE) {
+        chunk = blob;
+        blob = ByteString.EMPTY;
+      } else {
+        chunk = blob.substring(0, (int) DEFAULT_CHUNK_SIZE);
+        blob = blob.substring((int) DEFAULT_CHUNK_SIZE);
+      }
+      responseObserver.onNext(ReadResponse.newBuilder()
```

Comment:
FYI: This code loads the entire file into memory and buffers it in non-heap memory in Netty, so the memory doesn't show up in most Java memory profilers. In Bazel, we were running into the problem of running out of non-heap memory when we were uploading many large files concurrently. This could happen here as well. Unfortunately, gRPC only supports application-level flow control for the lower-level API (ServerCall), which we may have to implement here if this turns out to crash the server.